### ISSUE TITLE
Apache Phoenix support in Flyway 5.x

### ISSUE THREAD

**Issue by smmckay** [flyway/flyway#2112]
I see that Flyway 5.0 dropped support for Phoenix. I guess this was due to lack of interest in a fairly weird, niche DB. Looking at past issues it seems that Phoenix provides the necessary bits and pieces, although the Phoenix support in 4.x doesn't handle schemas.

**Comment by nycjay** [comment]
Yes - please add support back. I did not noticed it was removed when looking at the change logs, I must have missed it. We are running Flyway 4.x, and Phoenix.

**Comment by nycjay** [comment]
Any idea when this will be coming back? We are running 4.x, and were holding off on upgrading so we don't lose Phoenix support. If this won't get added in 6 we may need to fork and build ourself, maybe grabbing https://github.com/flyway/flyway/pull/2113

**Comment by axelfontaine** [comment]
Can you recommend a good Docker image to test this with tags for the various versions?

**Comment by nycjay** [comment]
Sorry, just noticed this comment. This looks like a suitable for testing Phoenix. 

https://hub.docker.com/r/boostport/hbase-phoenix-all-in-one/dockerfile

**Comment by Lyeeedar** [comment]
On investigation, we have hit a number of snags with reimplementing Phoenix support.

- The phoenix jdbc driver currently requires protobuf 2.5. We rely on protobuf 3.6 due to other dependencies, and do not wish to downgrade them to support a 7 year old library.
- Downgrading the protobuf version we rely on does allow us to bring in the phoenix-core dependency, and query the database. However as the library pulls in a huge number of dependencies we do not want this to be a default addition to flyway, instead being an optional driver. The phoenix-[version]-client driver should be the driver we need to include, but replacing the dependency added driver with the 'fat' one from the package does not work, failing on some internal reflection trying to create org.apache.phoenix.query.DefaultGuidePostsCacheFactory.
- Each version of the jdbc driver currently only supports a single phoenix/hbase version (5.0 phoenix with 2.0 hbase for example). As we need to compile against a single api that supports all versions users may intend to use, this provides us with a problem. There has been no statement on the side of phoenix about the volatility of the api, and as such we cannot commit ourselves to 10 years of support.

Obviously pull requests adding this support are welcome, but until the problems above are resolved we will not be committing to support Apache Phoenix.
